source/Octopus.Server.Client/Operations/RegisterMachineOperation.cs

```diff
@@ -225,28 +272,63 @@ async Task ValidateTenantTags(IOctopusSpaceAsyncRepository repository)
                 return;
 
             var tagSets = await repository.TagSets.FindAll().ConfigureAwait(false);
-            var missingTags = TenantTags.Where(tt => !tagSets.Any(ts => ts.Tags.Any(t => t.CanonicalTagName.Equals(tt, StringComparison.OrdinalIgnoreCase)))).ToList();
+            var missingTags = TenantTags.Where(tt =>
+                    !tagSets.Any(ts =>
+                        ts.Tags.Any(t => t.CanonicalTagName.Equals(tt, StringComparison.OrdinalIgnoreCase))))
+                .ToList();
 
             if (missingTags.Any())
-                throw new InvalidRegistrationArgumentsException(CouldNotFindMessage("tag", missingTags.ToArray()));
+                throw new InvalidRegistrationArgumentsException(CouldNotFindByNameMessage("tag",
+                    missingTags.ToArray()));
         }
 
         async Task<List<EnvironmentResource>> GetEnvironments(IOctopusSpaceAsyncRepository repository)
         {
-            var selectedEnvironments = new List<EnvironmentResource>();
-            foreach (var environmentName in EnvironmentNames)
+            List<EnvironmentResource> environments = new();
+            if (EnvironmentNames is not null && EnvironmentNames.Any())
             {
-                var environment = await repository.Environments.FindByName(environmentName).ConfigureAwait(false);
-                if (environment != null)
-                    selectedEnvironments.Add(environment);
+                var envsByName = await repository.Environments.FindByNames(EnvironmentNames).ConfigureAwait(false);
+                environments.AddRange(envsByName);
+
+                //if there are any missing environment names only, then we want to throw an exception and not check the missing names against slugs or ids
+                var missingByNameOnly = EnvironmentNames
+                    .Except(envsByName.Select(e => e.Name), StringComparer.OrdinalIgnoreCase).ToArray();
+
+                if (missingByNameOnly.Any())
+                    throw new InvalidRegistrationArgumentsException(
+                        CouldNotFindByNameMessage("environment", missingByNameOnly.ToArray()));
             }
 
-            var missing = EnvironmentNames.Except(selectedEnvironments.Select(e => e.Name), StringComparer.OrdinalIgnoreCase).ToList();
+            if (Environments is not null && Environments.Any())
+            {
+                var envsByName = await repository.Environments.FindByNames(Environments).ConfigureAwait(false);
+                environments.AddRange(envsByName);
 
-            if (missing.Any())
-                throw new InvalidRegistrationArgumentsException(CouldNotFindMessage("environment", missing.ToArray()));
+                //if there are any missing environment names only, then we want to throw an exception and not check the missing names against slugs or ids
```

Comment:
copy paste error

Reply:
fixed in [`073f520` (#845)](https://github.com/OctopusDeploy/OctopusClients/pull/845/commits/073f52085ff8429a3cc073f59d3148ed10ba9589)